scheduler.go

```diff
@@ -86,124 +86,148 @@ func (s *Scheduler) ChangeLocation(newLocation *time.Location) {
 }
 
 // scheduleNextRun Compute the instant when this Job should run next
-func (s *Scheduler) scheduleNextRun(j *Job) {
-	j.Lock()
-	defer j.Unlock()
+func (s *Scheduler) scheduleNextRun(job *Job) {
+	job.Lock()
+	defer job.Unlock()
 	now := s.time.Now(s.loc)
 
-	if j.startsImmediately {
-		j.nextRun = now
-		j.startsImmediately = false
+	if job.startsImmediately {
+		job.nextRun = now
+		job.startsImmediately = false
 		return
 	}
 
-	// delta represent the time slice used to calculate the next run
-	// it can be the last time ran by a job, or time.Now() if the job never ran
-	var delta time.Time
-	if j.neverRan() {
-		if !j.nextRun.IsZero() { // scheduled for future run, wait to run at least once
-			return
-		}
-		delta = now
-	} else {
-		delta = j.lastRun
+	if job.neverRan() && !job.nextRun.IsZero() {
+		return // scheduled for future run and should skip scheduling
 	}
+	job.lastRun = now
+
+	durationToNextRun := s.durationToNextRun(*job)
+	job.nextRun = job.lastRun.Add(durationToNextRun)
+}
 
-	switch j.unit {
+func (s *Scheduler) durationToNextRun(job Job) time.Duration {
+	lastRun := job.lastRun
+	var duration time.Duration
+	switch job.unit {
 	case seconds, minutes, hours:
-		j.nextRun = s.rescheduleDuration(j, delta)
+		duration = s.calculateDuration(job)
 	case days:
-		j.nextRun = s.rescheduleDay(j, delta)
+		duration = s.calculateDays(job, lastRun)
 	case weeks:
-		j.nextRun = s.rescheduleWeek(j, delta)
+		{
+			if job.scheduledWeekday != nil { // weekday selected, Every().Monday(), for example
+				duration = s.calculateWeekday(job, lastRun)
+			} else {
+				duration = s.calculateWeeks(job, lastRun)
+			}
+		}
 	case months:
-		j.nextRun = s.rescheduleMonth(j, delta)
+		duration = s.calculateMonths(job, lastRun)
+	}
+	return duration
+}
+
+func (s Scheduler) getJobLastRun(job Job) time.Time {
+	if job.neverRan() {
+		return s.time.Now(s.loc)
 	}
+	return job.lastRun
 }
 
-func (s *Scheduler) rescheduleMonth(j *Job, delta time.Time) time.Time {
-	if j.neverRan() { // calculate days to j.dayOfTheMonth
-		jobDay := time.Date(delta.Year(), delta.Month(), j.dayOfTheMonth, 0, 0, 0, 0, s.loc).Add(j.atTime)
-		daysDifference := int(math.Abs(delta.Sub(jobDay).Hours()) / 24)
-		nextRun := s.roundToMidnight(delta)
-		if jobDay.Before(delta) { // shouldn't run this month; schedule for next interval minus day difference
+func (s *Scheduler) calculateMonths(job Job, lastRun time.Time) time.Duration {
```

Comment:
the calculate*time* methods can be further simplified however I'm thinking about adding even more tests before moving on with any refactor